From 09536dbb2f2848792a99d0e30467e26ea917fca5 Mon Sep 17 00:00:00 2001
From: "ji, zhenlong z" <zhenlong.z.ji@intel.com>
Date: Thu, 17 Feb 2022 11:02:19 +0800
Subject: [PATCH] Set new umask value for mediaprovider app

Currently the umask value of mediaprovider app is 0077, that
means the other and group users cannot access the files created
by mediaprovider app. However, /sdcard/Pictures is a directory
shared by VM and Host and 9pfs is mounted on it. The user and
group of /sdcard/Pictures is system and ratio, so mediaprovier
app is a other user to this directory and its files, that means
mediaprovider app cannot access the files created by itself, this
is unreasonable. Set the umask value of mediaprovider app to 0, so
it can access the files created by itself.

Signed-off-by: ji, zhenlong z <zhenlong.z.ji@intel.com>
---
 src/com/android/providers/media/MediaProvider.java  | 2 ++
 src/com/android/providers/media/util/FileUtils.java | 3 ++-
 2 files changed, 4 insertions(+), 1 deletion(-)

diff --git a/src/com/android/providers/media/MediaProvider.java b/src/com/android/providers/media/MediaProvider.java
index 626bbb8e..46fe8e51 100644
--- a/src/com/android/providers/media/MediaProvider.java
+++ b/src/com/android/providers/media/MediaProvider.java
@@ -4825,6 +4825,8 @@ public class MediaProvider extends ContentProvider {
             } catch (FileNotFoundException ignored) {
             }
 
+            Os.umask(0);
+
             final File thumbDir = thumbFile.getParentFile();
             thumbDir.mkdirs();
 
diff --git a/src/com/android/providers/media/util/FileUtils.java b/src/com/android/providers/media/util/FileUtils.java
index 47c5bc14..e8f1e7ab 100644
--- a/src/com/android/providers/media/util/FileUtils.java
+++ b/src/com/android/providers/media/util/FileUtils.java
@@ -35,6 +35,7 @@ import static android.system.OsConstants.O_WRONLY;
 import static android.system.OsConstants.R_OK;
 import static android.system.OsConstants.S_IRWXG;
 import static android.system.OsConstants.S_IRWXU;
+import static android.system.OsConstants.S_IRWXO;
 import static android.system.OsConstants.W_OK;
 
 import static com.android.providers.media.util.DatabaseUtils.getAsBoolean;
@@ -99,7 +100,7 @@ public class FileUtils {
         final int posixFlags = translateModePfdToPosix(pfdFlags) | O_CLOEXEC | O_NOFOLLOW;
         try {
             final FileDescriptor fd = Os.open(file.getAbsolutePath(), posixFlags,
-                    S_IRWXU | S_IRWXG);
+                    S_IRWXU | S_IRWXG | S_IRWXO);
             try {
                 return ParcelFileDescriptor.dup(fd);
             } finally {
-- 
2.25.1

